rmw_connextdds_common/src/common/rmw_impl.cpp

```diff
@@ -1506,11 +1506,7 @@ RMW_Connext_Subscriber::loan_messages()
     "[%s] loaned messages: %lu",
     this->type_support->type_name(), this->loan_len)
 
-  if (this->loan_len > 0) {
-    return this->status_condition.set_data_available(true);
-  }
-
-  return RMW_RET_OK;
+  return this->status_condition.set_data_available(this->loan_len > 0);
```

Comment:
just to confirm, do you want to set the condition to `false` if `loan_len == 0`?

(I haven't checked the code well enough to see if that makes sense or not)

Reply:
Yes, that is the idea. `loan_len` keeps track of the samples loaned from the underlying DDS DataReader (via `DDS_DataReader_take()`), which the RMW caches and goes through them as they are taken by `rmw_take()`.

The flag on the status condition object is supposed to be true until the loan is consumed, returned, and a new one taken if available. As the flag is true, a `wait()` on the associated subscriber should return immediately.

(that's the goal, I suspect something is still not quite right in the implementation though...)

Reply:
> Yes, that is the idea. loan_len keeps track of the samples loaned from the underlying DDS DataReader (via DDS_DataReader_take()), which the RMW caches and goes through them as they are taken by rmw_take().

Out of curiosity, why is the rmw implementation caching the samples?

It also seems a bit confusing that the subscriber `has_data()`  method can trigger this, I would only expect the condition to be updated from the "on data listener" and reset from the "take_*" methods.

Reply:
> Out of curiosity, why is the rmw implementation caching the samples?

I found it to be a good way to keep track of new available samples in a reader's cache while (re)implementing event handling to support ROS 2 WaitSets.

Neither Pro nor Micro offer an API to query a reader for whether there are new samples in the cache, which is needed to properly manage the `triggered_data` flag inside the RMW, especially if implementing on top of DDS WaitSets, since the DATA_AVAILABLE status is reset at the first `take()`/`read()`, and since ROS 2 doesn't generally consume all available samples once a subscriber is returned by a `wait()`, it is necessary to keep track of this state in the RMW. One could implement it by querying the reader cache with a `read()`, but I thought at that point, it would be best to just `take()` everything available out of the cache, then use the length of the result (`loan_len`) to keep track of whether there are new samples to return (and have `triggered_data` set accordingly).

This `take()` should be relatively inexpensive since there isn't any "deep copy" nor deserialization occurring, only pointers being copied into what are pre-allocated sequences (eventually, since they are allocated as needed but then cached and reused).

> It also seems a bit confusing that the subscriber has_data() method can trigger this, I would only expect the condition to be updated from the "on data listener" and reset from the "take_*" methods.

Querying and caching the reader in one move seems better than accessing the cache multiple times, once to peek if there are samples (to support `has_data()`), and then to actually `take()` samples (to support the `rmw_take_*` methods).

Since the `take()` should be inexpensive, I think it should be fine, although it may be revised if we see some problems.

The samples are only cached as pointers to their serialized buffers, and they are only deserialized "opportunistically" within the `rmw_take_*` operations (see `RMW_Connext_Subscriber::take_next()`).